### ISSUE TITLE
Support reloading TlsCertificate files

### ISSUE THREAD

**Issue by howardjohn** [envoyproxy/envoy#10387]
**Issue Template**

*Title*: Support reloading TlsCertificate files

*Description*:

Currently, there are a few ways to get certificates reloaded (that I know of!):

* Hot restart - this works but requires a wrapper that is watching the certs and hot restarting envoy when they reload. This adds complexity since you need to create/maintain a wrapper, which also implies your own docker image. It also has a run time cost when hot restarting.
* SDS - this was, as far as I can tell, intended for solving this purpose. However, it involves quite a bit of complexity for the control plane, and another moving part for Envoy.
* [File based SDS](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#filesystem-subscriptions) - I am not sure if this is intended for production usage or testing/debugging? Regardless, this also has the issue that it requires a custom format. This means that in order to use existing tooling that is provisioning certificates, we need another wrapper to watch those files and convert them to SDS shaped files.

As someone not familiar with Envoy internals, it seems like the ideal scenario would be for Envoy to just watch the certificates in the first place. When they are reloaded on the disk, they would be reloaded internally.

From the outside this seems like it should be feasible as
* SDS is already hot reloading certs, so that code exists somewhere
* Files are already watched for other parts of the code as well

Is this something that would be considered useful or feasible to implement? From looking at a few past issues it seems the answer has been "use SDS", but as discussed above I don't think that is the best solution (from a user perspective).

**Comment by howardjohn** [comment]
This would also help a lot if you were just using a simple static configuration

**Comment by htuch** [comment]
I think the most reasonable thing would be to use the SDS implementation and config as is, but allow for it to do a filesystem watch on the `DataSource` for the cert. This probably isn't too hard to implement given the SDS subsystem takes care of lifecycle issues already.

**Comment by mattklein123** [comment]
> I think the most reasonable thing would be to use the SDS implementation and config as is, but allow for it to do a filesystem watch on the DataSource for the cert. This probably isn't too hard to implement given the SDS subsystem takes care of lifecycle issues already.

+1 I was going to suggest something similar.

**Comment by ggreenway** [comment]
@htuch or @howardjohn, is this resolved already, either by #13721, or by #10163?

**Comment by howardjohn** [comment]
I think those are basically file-based SDS? which has the issues listed here:
> File based SDS - I am not sure if this is intended for production usage or testing/debugging? Regardless, this also has the issue that it requires a custom format. This means that in order to use existing tooling that is provisioning certificates, we need another wrapper to watch those files and convert them to SDS shaped files.
if I understand right.

For concrete use case, this would prevent simply mounting kubernetes Secrets created by tools like cert-manager

**Comment by ggreenway** [comment]
File-based SDS can point to another two files for the cert and key. So yes an SDS file needs to be created, but it could be a static file, and result in watches on the cert and key files. At least that's my read of what's been implemented.

**Comment by htuch** [comment]
@ggreenway correct, and since #13721 I think we in a position to say this works in a sensible way that is compatible with things like k8s rotations.

**Comment by ggreenway** [comment]
Should we close this? The only other thing that may be nice-to-have is to have a flag to automatically setup the file-watching without needing an SDS file to set it up, eg have a bool in the settings that when true, envoy acts like it read in a filesystem-based SDS response that points to the paths for the cert/key and watches them for changes. But that doesn't add any new functionality; it just makes it slightly easier to setup the existing.

**Comment by htuch** [comment]
Yeah, we could definitely do something like that if it's a major pain point. As long as we don't build two FS watch mechanisms I'm all good.

**Comment by karthikbox** [comment]
A huge +1 to Howard's [comment](https://github.com/envoyproxy/envoy/issues/10387#issuecomment-840750718) . Not requiring the SDS file to be setup would be awesome. This is definitely a huge pain point. 

Just so that i understood correctly - the proposal above is to setup a watch for [TlsCertificate](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-tlscertificate) (similarly for [CertificateValidationContext](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-certificatevalidationcontext) even when they're not delivered by SDS, correct?

Also, why not make this the default behavior?
